### PR TITLE
 core: (irdl) Add support for regions in optional groups to declarative assembly format

### PR DESCRIPTION
**Pull request by kimxworrall**
Added support for regions in optional groups and tests for this.

### CODE REVIEW

File: tests/irdl/test_declarative_assembly_format.py [hunk 1]
```diff
@@ -1312,23 +1304,66 @@ class OptionalRegionOp(IRDLOperation):
...
+            'test.optional_region_group keyword {\n  "test.op"() : () -> ()\n}',
+            '"test.optional_region_group"() ({ "test.op"() : () -> ()}) : () -> ()',
+        ),
+    ],
+)
+def test_optional_groups_regions(format: str, program: str, generic_program: str):
+    """Test the parsing of optional regions in an optional group"""
+
+    @irdl_op_definition
+    class OptionalRegionOp(IRDLOperation):  # pyright: ignore[reportUnusedClass]
```
Comment (superlopuh) [inline tests/irdl/test_declarative_assembly_format.py:1352]: Isn't it used below?
```suggestion
    class OptionalRegionOp(IRDLOperation):
```

File: xdsl/irdl/declarative_assembly_format_parser.py [hunk 2]
```diff
@@ -187,12 +188,25 @@ def verify_directives(self, elements: list[FormatDirective]):
...
-                    isinstance(a, VariadicLikeTypeDirective)
-                    or isinstance(b, VariadicLikeTypeDirective)
-                ):
+                case VariadicLikeVariable(), VariadicLikeVariable():
+                    if not (
+                        isinstance(a, VariadicLikeTypeDirective)
+                        or isinstance(b, VariadicLikeTypeDirective)
+                        or isinstance(a, RegionDirective)
+                        or isinstance(b, RegionDirective)
+                    ):
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format_parser.py:197]: Should be equivalent and marginally faster
```suggestion
                    if not (
                        isinstance(a, RegionDirective, VariadicLikeTypeDirective)
                        or isinstance(b, RegionDirective, VariadicLikeTypeDirective)
                    ):
```

**Review by superlopuh — APPROVED** [review]

**Comment by alexarice** [comment]
Perhaps you and @superlopuh have discussed this in person but would it be possible to describe what wasn't working before/brief summary of what you've changed, as there seems to be lots of changes in the "files changed section"

File: xdsl/irdl/declarative_assembly_format_parser.py [hunk 3]
```diff
@@ -187,12 +188,23 @@ def verify_directives(self, elements: list[FormatDirective]):
                     self.raise_error(
                         "A variadic type directive cannot be followed by another variadic type directive."
                     )
-                case VariadicLikeVariable(), VariadicLikeVariable() if not (
-                    isinstance(a, VariadicLikeTypeDirective)
-                    or isinstance(b, VariadicLikeTypeDirective)
-                ):
+                case VariadicLikeVariable(), VariadicLikeVariable():
+                    if not (
+                        isinstance(a, RegionDirective, VariadicLikeTypeDirective)
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format_parser.py:193]: My bad
```suggestion
                        isinstance(a, RegionDirective | VariadicLikeTypeDirective)
```

**Comment by kimxworrall** [comment]
This commit adds a RegionDirective that all RegionVariables inherit from. When sequences of variables in declarative assembly format are checked, if two consecutive variables are optional (either two optional operands or two optional regions in a row, but not one optional region and one optional operand), an error is raised to inidicate ambiguity.

test_optional_groups_regions is modified to catch this error, and two other tests are made cleaner as hang overs from the previous pr on this branch.

**Review by superlopuh — APPROVED** [review]

**Comment by alexarice** [comment]
Is this ready to merge?

**Comment by superlopuh** [comment]
Yeah seems ready to me, feel free to hit the merge button at your convenience :) @kimxworrall

**Pull request merged**